arrow/src/datatypes/field.rs

```diff
@@ -107,6 +107,47 @@ impl Field {
         self.nullable
     }
 
+    /// Returns a (flattened) vector containing all fields contained within this field (including it self)
+    pub(crate) fn fields(&self) -> Vec<&Field> {
+        let mut collected_fields = vec![self];
+        match &self.data_type {
+            DataType::Struct(fields) | DataType::Union(fields) => collected_fields
+                .append(&mut fields.iter().map(|f| f.fields()).flatten().collect()),
```

Comment:
I think you can save an intermediate `collect` here by doing something like:

```suggestion
            DataType::Struct(fields) | DataType::Union(fields) => collected_fields
                .extend(fields.iter().map(|f| f.fields()).flatten()),
```

Probably doesn't make any practical difference, mostly a stylistic difference